src/main/java/com/google/api/generator/gapic/composer/ServiceClientClassComposer.java

```diff
@@ -1667,4 +1704,49 @@ private static boolean isProtoEmptyType(TypeNode type) {
     return type.reference().pakkage().equals("com.google.protobuf")
         && type.reference().name().equals("Empty");
   }
+
+  private static void updateGapicMetadata(
+      GapicContext context, String protoPackage, String javaPackage) {
+    context.updateGapicMetadata(
+        context
+            .gapicMetadata()
+            .toBuilder()
+            .setProtoPackage(protoPackage)
+            .setLibraryPackage(javaPackage)
+            .build());
+  }
+
+  private static void updateGapicMetadata(
+      GapicContext context,
+      Service service,
+      String clientClassName,
+      Map<String, List<String>> grpcRpcToJavaMethodNames) {
+    GapicMetadata.Builder metadataBuilder = context.gapicMetadata().toBuilder();
+    metadataBuilder =
+        metadataBuilder
+            .setProtoPackage(service.protoPakkage())
+            .setLibraryPackage(service.pakkage());
+
+    GapicMetadata.ServiceAsClient.Builder serviceClientProtoBuilder =
+        GapicMetadata.ServiceAsClient.newBuilder().setLibraryClient(clientClassName);
+
+    // Sort for deterministic tests.
+    List<String> sortedRpcNames = new ArrayList<>(grpcRpcToJavaMethodNames.keySet());
+    Collections.sort(sortedRpcNames);
+    for (String rpcName : sortedRpcNames) {
+      GapicMetadata.MethodList methodList =
+          GapicMetadata.MethodList.newBuilder()
+              .addAllMethods(grpcRpcToJavaMethodNames.get(rpcName))
+              .build();
+      serviceClientProtoBuilder.putRpcs(rpcName, methodList);
+    }
+
+    metadataBuilder =
```

Comment:
I'm curious: I see you use the pattern `fooBuilder = fooBuilder.doSomething().setSomething()`. Is the assignment redundant? I thought the builder methods were mutating, rather than functional methods returning a new copy of the object with the changes requested.

Reply:
No, the original builder doesn't mutate -  AFAIK AutoValue classes are all immutable.

Reply:
Oh, weird. In the monolith, builders were used as though their setters mutate. My mental model had been that AutoValue classes are immutable but AutoValue.Builder clases were mutable; I'll need to figure this out. Is there more than one AutoValue impl around?